sphinx_needs/views.py

```diff
@@ -0,0 +1,225 @@
+            self._maybe_len = len(self._needs)
+        return self._maybe_len
+
+    def __bool__(self) -> bool:
+        try:
+            next(iter(self._needs))
+        except StopIteration:
+            return False
+        return True
+
+    def to_list(self) -> NeedsListView:
+        return NeedsListView(_needs=self)
+
+    def to_list_with_parts(self) -> NeedsAndPartsListView:
+        return NeedsAndPartsListView(_needs=self)
+
+    def filter_ids(self, values: list[str]) -> NeedsView:
+        """Create new view with needs filtered by the ``id`` field."""
+        return self._copy_filtered(values)
+
+    def filter_is_external(self, value: bool) -> NeedsView:
+        """Create new view with needs filtered by the ``is_external`` field."""
+        return self._copy_filtered(self._indexes().is_external.get(value, []))
+
+    def filter_types(self, values: list[str], or_type_names: bool = False) -> NeedsView:
+        """Create new view with needs filtered by the ``type`` field.
+
+        :param values: List of types to filter by.
+        :param or_type_names: If True, filter by both ``type`` and ``type_name`` field
+        """
+        if or_type_names:
+            return self._copy_filtered(
+                i
+                for value in values
+                for i in chain(
+                    self._indexes().types.get(value, []),
+                    self._indexes().type_names.get(value, []),
+                )
+            )
+        return self._copy_filtered(
+            i for value in values for i in self._indexes().types.get(value, [])
+        )
+
+    def filter_statuses(self, values: list[str]) -> NeedsView:
+        """Create new view with needs filtered by the ``status`` field."""
+        return self._copy_filtered(
+            i for value in values for i in self._indexes().status.get(value, [])
+        )
+
+    def filter_tags(self, values: list[str]) -> NeedsView:
+        """Create new view with needs filtered by the ``tags`` field."""
+        return self._copy_filtered(
+            i for value in values for i in self._indexes().tags.get(value, [])
+        )
+
+
+class NeedsListView:
+    """A read-only view of needs,
+    after resolution (e.g. back links have been computed etc)
+    """
+
+    __slots__ = ("_needs",)
+
+    def __init__(self, *, _needs: NeedsView):
+        """ "Initialize a new view of needs.
+
+        .. important:: This class is not meant to be instantiated by users,
+            a singleton instance is managed by sphinx-needs.
+
+            Instead, create from a NeedsView instance.
+        """
+        self._needs = _needs
+
+    def __bool__(self) -> bool:
+        return bool(self._needs)
+
+    def __iter__(self) -> Iterator[NeedsInfoType]:
+        return iter(self._needs.values())
+
+    def __len__(self) -> int:
+        return len(self._needs)
+
+    def to_map_view(self) -> NeedsView:
+        return self._needs
+
+
+class NeedsAndPartsListView:
```

Comment:
How would I get all parts of a certain need?

Reply:
well it would be via; https://github.com/useblocks/sphinx-needs/blob/dd79aa3f18b3f8ee7a3a7ab8d7bb30b77dc5232b/sphinx_needs/roles/need_part.py#L44

although to note, this is not currently part of the "public API" (and never has been)

to note also, this `NeedsAndPartsListView` view is added separately so as not to be back-breaking for current user code (see https://github.com/useblocks/sphinx-needs/issues/1264),
if it was not for this, I would probably have just added an iter method to `NeedsView`